### PR TITLE
Add Watch of Velero Backup and remove Backup controller

### PR DESCRIPTION
**Pull request by mpryc**
Change which removes Backup Controller and adds the Watch of the Backup object within NonAdminBackup Controller.

Added two main predicates with 'upper' logic within composite predicate to easily distinguish between predicate for NonAdminBackup and the Backup one.

The reconcile loop updates the Status and the Spec of the NonAdminBackup from the Velero Backup object. The Spec in the Velero Backup is updated by the Velero, so we use similar logic to update our NonAdminBackup one.

Fixes https://github.com/migtools/oadp-non-admin/issues/11

### CODE REVIEW

File: cmd/main.go
Comment (mateusoliveira43) [inline cmd/main.go:1]: does this need to reverted as well https://github.com/migtools/oadp-non-admin/commit/5ff6f33855b8d8876dff48579fbb6f02cda23f7e#diff-c444f711e9191b53952edb65bfd8c644419fc7695c62611dc0fb304b4fb197d6R54 ?

**Comment by mpryc** [comment]
> Also, undo changes from PROJECT file [5ff6f33#diff-d1fb402f1269ba7396dca9cbd03b44ac055874669bce1d0f239fa2f54ea3584bR20](https://github.com/migtools/oadp-non-admin/commit/5ff6f33855b8d8876dff48579fbb6f02cda23f7e#diff-d1fb402f1269ba7396dca9cbd03b44ac055874669bce1d0f239fa2f54ea3584bR20)

We can't because we will get error:

```
ERROR	controller-runtime.source.EventHandler	kind must be registered to the Scheme	{"error": "no kind is registered for the type v1.Backup in scheme \"pkg/runtime/scheme.go:100\""
```

File: internal/controller/velerobackup_handler.go
Comment (mateusoliveira43) [inline internal/controller/velerobackup_handler.go:1]: why this file? what is the difference between this and predicate?

Comment (mpryc) [reply internal/controller/velerobackup_handler.go:1]: Yes. Handler handles events which are let by predicates. They are together but separate:

```
EVENT ---> PREDICATE (true) ----> HANDLER ----> DO something

EVENT ----> PREDICATE (false) ---> Do nothing
```

File: internal/controller/velerobackup_handler.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+func (h *VeleroBackupHandler) Create(ctx context.Context, evt event.CreateEvent, q workqueue.RateLimitingInterface) {
+	nameSpace := evt.Object.GetNamespace()
+	name := evt.Object.GetName()
+	log := getVeleroBackupHandlerLogger(ctx, name, nameSpace)
+	log.V(1).Info("Received Create VeleroBackupHandler")
+}
+
+func (h *VeleroBackupHandler) Update(ctx context.Context, evt event.UpdateEvent, q workqueue.RateLimitingInterface) {
+	nameSpace := evt.ObjectNew.GetNamespace()
+	name := evt.ObjectOld.GetName()
```
Comment (mateusoliveira43) [inline internal/controller/velerobackup_handler.go:35]: why old here?

File: internal/controller/nonadminbackup_predicate.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+		nameSpace := evt.Object.GetNamespace()
+		name := evt.Object.GetName()
+		log := getNonAdminBackupPredicateLogger(ctx, name, nameSpace)
+		log.V(1).Info("Received Delete NonAdminBackupPredicate")
+		return true
+	}
+	return false
+}
+
+func (predicate NonAdminBackupPredicate) Generic(ctx context.Context, evt event.GenericEvent) bool {
```
Comment (mateusoliveira43) [inline internal/controller/nonadminbackup_predicate.go:58]: generic would be the default of switch case?

File: internal/controller/nonadminbackup_controller.go [hunk 3]
```diff
@@ -142,5 +148,13 @@ func (r *NonAdminBackupReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 func (r *NonAdminBackupReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&nacv1alpha1.NonAdminBackup{}).
+		Watches(&velerov1api.Backup{}, &VeleroBackupHandler{}).
```
Comment (mateusoliveira43) [inline internal/controller/nonadminbackup_controller.go:151]: how we know the difference between update and patch?

File: internal/controller/velerobackup_predicate.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+	return nameSpace == veleroBackupPredicate.OadpVeleroNamespace
+}
+
+func (veleroBackupPredicate VeleroBackupPredicate) Delete(ctx context.Context, evt event.DeleteEvent) bool {
+	// NonAdminBackup should not care about VeleroBackup delete events
+	return false
+}
+
+func (veleroBackupPredicate VeleroBackupPredicate) Generic(ctx context.Context, evt event.GenericEvent) bool {
```
[the remark this answers is not in the document]
Comment (mpryc) [reply internal/controller/velerobackup_predicate.go:67]: We need as we are implementing interface.

File: internal/controller/nonadminbackup_predicate.go (hunk 2 above)
Comment (mpryc) [reply internal/controller/nonadminbackup_predicate.go:58]: Here you have more about those 4 predicates, it's interface which we need to implement:
https://github.com/kubernetes-sigs/controller-runtime/blob/v0.17.2/pkg/predicate/predicate.go#L32

File: internal/controller/composite_predicate.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+	return p.VeleroBackupPredicate.Delete(p.Context, evt)
+}
+
+func (p CompositePredicate) Generic(evt event.GenericEvent) bool {
+	// If NonAdminBackupPredicate returns true, ignore VeleroBackupPredicate
+	if p.NonAdminBackupPredicate.Generic(p.Context, evt) {
+		return true
+	}
+	// Otherwise, apply VeleroBackupPredicate
+	return p.VeleroBackupPredicate.Generic(p.Context, evt)
```
[the remark this answers is not in the document]
Comment (mpryc) [reply internal/controller/composite_predicate.go:50]: We could but then it's harder to put some clear comments.

[the remark this answers is not in the document]
Comment (mpryc) [reply internal/controller/composite_predicate.go:50]: Plus there may be other conditions for predicate for each of the objects, so having each block for each "type" makes sense imo.

File: internal/controller/nonadminbackup_controller.go (hunk 3 above)
Comment (kaovilai) [reply internal/controller/nonadminbackup_controller.go:151]: For events, everything is called update including patch :)

File: internal/controller/velerobackup_predicate.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+func (veleroBackupPredicate VeleroBackupPredicate) Create(ctx context.Context, evt event.CreateEvent) bool {
+	nameSpace := evt.Object.GetNamespace()
+	name := evt.Object.GetName()
+	log := getBackupPredicateLogger(ctx, name, nameSpace)
+	log.V(1).Info("Received Create VeleroBackupPredicate")
+
+	if nameSpace != veleroBackupPredicate.OadpVeleroNamespace {
+		return false
+	}
```
Comment (kaovilai) [inline internal/controller/velerobackup_predicate.go:40]:
```suggestion
	nameSpace := evt.Object.GetNamespace()
	if nameSpace != veleroBackupPredicate.OadpVeleroNamespace {
		return false
	}

	name := evt.Object.GetName()
	log := getBackupPredicateLogger(ctx, name, nameSpace)
	log.V(1).Info("Received Create VeleroBackupPredicate")
```

for performance, let's not call getName, getLogger, log if the namespace is different.

**Review by kaovilai — CHANGES_REQUESTED** [review]
predicates should be very performant cause there can be many hundreds of these at any given time that need filtering out.

File: internal/controller/velerobackup_handler.go
Comment (mateusoliveira43) [reply internal/controller/velerobackup_handler.go:1]: controller runtime calls it internally event handler https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.17.2/pkg/handler#EventHandler and predicate https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.17.2/pkg/predicate#Predicate

**Comment by mpryc** [comment]
@mateusoliveira43 could we merge this now?

**Review by mateusoliveira43 — APPROVED** [review]
lets merge to unblock PRs, we can analyze code more in depth next week

**Pull request merged**